docs/cplusplus-style-guide.md

```diff
@@ -0,0 +1,372 @@
+
+
+## Formatting
+
+* Make use of the `apply_style` script provided for formatting.
+* Lines should be no longer than 100 characters.
+* Use 2 spaces to indent, not tabs.
+* Curly brackets `{...}` should be on their own line.
+* There should be a space after conditionals, i.e. `if (test)`
+* No spaces inside `(parenthesis)` or `<template>` unless it improves readability.
+* No spaces around arrow: `this->fnc();`.
+* No trailing whitespace.
+* Align attributes by group in classes, and initialisers.
+* Variables should be initialised with `=` if primitive, or braced initialisation with parentheses if required.
+* Initialiser lists should be initialised with braces by default.
+* Use SQL style initialiser for lists and inheritance lists.
+
+```c++
+Foo::Foo()
+  : attribute1_{value1}
+  , attribute2_{value2}
+{}
+```
+
+
+## Typing
+
+* Use STL types wherever possible.
+* Prefer `(u)int(16/32/64)_t` and `std::size_t`.
+* Use `static_cast`, `reinterpret_cast`, `dynamic_cast`, etc., rather than C-style casts.
+* Avoid `auto` except when the return type is clear.
+* Avoid use of RTTI.
+
+* `using` is preferable to `typedef` for type alias definition.
+* Make `using` private unless you intend to expose it.
+* `namespace {.<template T>` only allowed if `typedef`/`using` indicates the type.
+
+* If a value is `const`, declare it so.
+* Use `int const &i` over `const int &i`.
+
+* Prefer `enum class` over regular `enum`.
+
+## Namespaces
+
+* Do not do `using namespace foo` or `namespace baz = ::foo::bar::baz` in `.hpp` files.
+* Do not do `using namespace foo` in the top `fetch` namespace, as it pollutes all of it. Instead, do the following:
+
+```c++
+namespace fetch
+{
+  namespace details
+  {
+    struct FetchImpl
+    {
+      /* ... */
+    };
+
+    void HelperFunction(FetchImpl &s)
+    {
+      /* ... */
+    }
+    };  // namespace details
+
+    class MyFetchClass
+    {
+
+    public:
+      void DoSomeThing()
+      {
+        // This improves readability
+        using namespace details;
+
+        FetchImpl f;
+        HelperFunction(f);
+      }
+      // ...
+    };
+  };
+};
+```
+
+
+## Structs
+
+Use structs for Plain Old Data (POD) and, in some cases, template meta-programming. Use `classes` for everything else.
+
+``` c++
+struct Person
+{
+  std::string name;
+  int         age;
+  // ...
+};
+```
+
+## Classes
+
+* Initialise class members in the following order of preference:
+  1. Inline initialisation. This ensures, for instance, that the default constructor does not produce uninitialised memory.
+  2. Constructor initialiser list. Member variables must be listed in the order they appear.
+  3. Constructor body.
+
+In most cases, a combination of the above will be used. For example:
+
+``` c++
+class MyClass
+{
+public:
+  // Inline initialisation
+  MyClass() = default;
+
+  // Initialisation list for size, constructor body for pointer
+  explicit MyClass(std::size_t n)
+    : size_(n)
+  {
+    pointer_ = new int[n];
```

Comment:
Isn't this assigning an array of ints to a int * here?

Reply:
Unfortunate given that we discourage the usage of raw pointers. I will change this just to illustrate what the different types of initialisation look like.

Reply:
👍